### PR TITLE
IIR refactor VI

### PR DESCRIPTION
**Pull request by cosunae**
Final IIR refactor that implements all the comments in PR

### CODE REVIEW

**Comment by cosunae** [comment]
launch pipeline

**Comment by cosunae** [comment]
launch pipeline

**Comment by cosunae** [comment]
launch pipeline

File: src/dawn/IIR/Stage.cpp [hunk 1]
```diff
@@ -12,46 +12,55 @@
...
 
-bool Stage::hasSingleDoMethod() const { return (DoMethods_.size() == 1); }
+  cloneStage->cloneChildrenFrom(*this);
+  return cloneStage;
+}
 
 DoMethod& Stage::getSingleDoMethod() {
   DAWN_ASSERT_MSG(hasSingleDoMethod(), "stage contains multiple Do-Methods");
-  return *DoMethods_.front();
+  return *(getChildren().front());
```
Comment (twicki) [inline src/dawn/IIR/Stage.cpp]: here we want `children_`, right?

File: src/dawn/IIR/Stage.cpp [hunk 2]
```diff
@@ -12,46 +12,55 @@
...
 }
 
 const DoMethod& Stage::getSingleDoMethod() const {
   DAWN_ASSERT_MSG(hasSingleDoMethod(), "stage contains multiple Do-Methods");
-  return *DoMethods_.front();
+  return *(getChildren().front());
```
Comment (twicki) [inline src/dawn/IIR/Stage.cpp]: here we want `children_`, right?

File: src/dawn/IIR/Stage.cpp [hunk 3]
```diff
@@ -12,46 +12,55 @@
...
 }
 
+bool Stage::hasSingleDoMethod() const { return (children_.size() == 1); }
+
 boost::optional<Interval>
 Stage::computeEnclosingAccessInterval(const int accessID, const bool mergeWithDoInterval) const {
   boost::optional<Interval> interval;
-  for(auto const& doMethod : DoMethods_) {
+  for(auto const& doMethod : getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/Stage.cpp]: here we want `children_`, right?

File: src/dawn/IIR/Stage.cpp [hunk 4]
```diff
@@ -67,47 +76,57 @@ Stage::computeEnclosingAccessInterval(const int accessID, const bool mergeWithDo
...
+  std::transform(childrenBegin(), childrenEnd(), std::back_inserter(intervals),
+                 [&](const DoMethodSmartPtr_t& doMethod) { return doMethod->getInterval(); });
   return intervals;
 }
 
 Interval Stage::getEnclosingInterval() const {
-  Interval interval = DoMethods_.front()->getInterval();
-  for(int i = 1; i < DoMethods_.size(); ++i)
-    interval.merge(DoMethods_[i]->getInterval());
+  Interval interval = getChildren().front()->getInterval();
```
Comment (twicki) [inline src/dawn/IIR/Stage.cpp]: search and replace in this file...

File: src/dawn/IIR/MultiStage.cpp [hunk 5]
```diff
@@ -268,52 +282,49 @@ boost::optional<Interval> MultiStage::getEnclosingAccessIntervalTemporaries() co
...
 
-    stage.update();
+    stage.update(NodeUpdateType::levelAndTreeAbove);
   }
 }
 
 bool MultiStage::isEmptyOrNullStmt() const {
-  for(auto stageIt = getStages().begin(); stageIt != getStages().end(); ++stageIt) {
-    if(!(*stageIt)->isEmptyOrNullStmt())
+  for(const auto& stage : getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/MultiStage.cpp]: here we want `children_`, right?

File: src/dawn/IIR/IIRNode.h [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+  template <typename TChildSmartPtr>
+  void
+  setChildrenParent(typename std::enable_if<!is_child_void<TChildSmartPtr>::value>::type* = 0) {
+    if(!parent_)
+      return;
+
+    const std::unique_ptr<NodeType>& thisNodeSmartPtr =
+        (*parent_)->getChildSmartPtr(static_cast<NodeType*>(this));
+
+    for(const auto& child : getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h]: here we want `children_`, right?

File: src/dawn/IIR/IIRNode.h [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+  checkTreeConsistencyImpl(typename std::enable_if<std::is_void<TChild>::value>::type* = 0) const {
+    return true;
+  }
+
+  template <typename TChild>
+  bool
+  checkTreeConsistencyImpl(typename std::enable_if<!std::is_void<TChild>::value>::type* = 0) const {
+    PROTECT_TEMPLATE(TChild, Child)
+
+    for(const auto& child : getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h]: here we want `children_`, right?

File: src/dawn/IIR/IIRNode.h [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+    inputChild
+        ->template setChildrenParent<typename getChildType<std::unique_ptr<Child>>::type::type>();
+
+    updateFromChildrenRec<NodeType>();
+  }
+
+  /// @brief print the tree of pointers (for debugging)
+  template <typename TChild>
+  void printTreeImpl(typename std::enable_if<!std::is_void<TChild>::value>::type* = 0) {
+    for(const auto& child : getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h]: here we want `children_`, right?

File: src/dawn/IIR/DoMethod.cpp [hunk 9]
```diff
@@ -12,44 +12,46 @@
...
-std::shared_ptr<DependencyGraphAccesses>& DoMethod::getDependencyGraph() {
-  return dependencyGraph_;
+  derivedInfo_.dependencyGraph_ = DG;
 }
 
 boost::optional<Extents> DoMethod::computeMaximumExtents(const int accessID) const {
   boost::optional<Extents> extents;
 
-  for(auto& stmtAccess : getStatementAccessesPairs()) {
+  for(auto& stmtAccess : getChildren()) {
```
Comment (twicki) [inline src/dawn/IIR/DoMethod.cpp]: here we want `children_`, right?

File: src/dawn/IIR/IIRNode.h [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+
+  /// @brief set the parent pointer of a child of this node
+  /// @param child child node of this node, for which the parent pointer will be set
+  template <typename TParent, typename TChild>
+  void setChildParent(
+      const std::unique_ptr<TChild>& child,
+      typename std::enable_if<!std::is_void<TParent>::value &&
+                              !is_child_void<std::unique_ptr<TChild>>::value>::type* = 0) {
+
+    static_assert(std::is_same<TParent, Parent>::value,
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h]: user the macro defined in line 31

File: src/dawn/IIR/IIRNode.h [hunk 11]
```diff
@@ -0,0 +1,616 @@
...
+  }
+
+  template <typename TChild>
+  void cloneChildrenImpl(const IIRNode& other,
+                         typename std::enable_if<std::is_void<TChild>::value>::type* = 0) {}
+
+  template <typename TChild>
+  void cloneChildrenImpl(const IIRNode& other,
+                         typename std::enable_if<!std::is_void<TChild>::value>::type* = 0) {
+    static_assert(std::is_same<TChild, Child>::value,
```
Comment (twicki) [inline src/dawn/IIR/IIRNode.h]: user the macro defined in line 31

File: src/dawn/IIR/IIRNodeIterator.h [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    auto it = IIRNodeIterator<RootNode, LeafNode>(&root_);
+    it.setToEnd();
+    return it;
+  }
+};
+
+} // namespace iir
+
+template <typename Leaf, typename RootNode>
+iir::IIRRange<RootNode, Leaf> iterateIIROver(const RootNode& root) {
```
Comment (twicki) [inline src/dawn/IIR/IIRNodeIterator.h]: are we sticking with this naming? My initial comment was:
 
I'm not sure about the name here, evtl use something that tells that you iterate over the children('s childern) of the current node? If I read `iteateIIROver<Stages>(*this)` this reads as iterating over all stages. I don't know the best solution yet

File: src/dawn/IIR/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+##===------------------------------------------------------------------------------------------===##
+
+yoda_add_library(
+  NAME DawnIIR
+  SOURCES 
+          Accesses.cpp  
+          Accesses.h
+          BlockStatements.cpp
+          BlockStatements.h
+          Cache.cpp
```
Comment (twicki) [inline src/dawn/IIR/CMakeLists.txt]: how big should the subproject be? does it make sense to have everything in one folder or should we aim for a separation of "tree" and IIR-Utils?

File: src/dawn/IIR/IntervalAlgorithms.cpp [hunk 14]
```diff
@@ -12,10 +12,11 @@
 //
 //===------------------------------------------------------------------------------------------===//
 
-#include "dawn/Optimizer/IntervalAlgorithms.h"
+#include "dawn/IIR/IntervalAlgorithms.h"
 
 namespace dawn {
-MultiInterval substract(const Interval& int1, const Interval& int2) {
+namespace iir {
+MultiInterval substract(const iir::Interval& int1, const iir::Interval& int2) {
```
Comment (twicki) [inline src/dawn/IIR/IntervalAlgorithms.cpp]: remove the iir namespace -> go to `const Interval& int1`

File: src/dawn/IIR/IntervalAlgorithms.h [hunk 15]
```diff
@@ -13,19 +13,20 @@
...
+#include "dawn/IIR/Cache.h"
+#include "dawn/IIR/Interval.h"
+#include "dawn/IIR/MultiInterval.h"
 
 namespace dawn {
+namespace iir {
 
 /// @brief substracts (relative complement) two intervals, generating as a result a multiinterval
-MultiInterval substract(const Interval& int1, const Interval& int2);
+MultiInterval substract(const iir::Interval& int1, const iir::Interval& int2);
```
Comment (twicki) [inline src/dawn/IIR/IntervalAlgorithms.h]: again, namespace obsolete

File: src/dawn/IIR/MultiInterval.cpp [hunk 16]
```diff
@@ -25,7 +26,7 @@ std::ostream& operator<<(std::ostream& os, const MultiInterval& multiInterval) {
   return os;
 }
 
-MultiInterval::MultiInterval(std::initializer_list<Interval> const& intervals)
+MultiInterval::MultiInterval(std::initializer_list<iir::Interval> const& intervals)
```
Comment (twicki) [inline src/dawn/IIR/MultiInterval.cpp]: again, namespace obsolete

File: src/dawn/IIR/MultiInterval.h [hunk 17]
```diff
@@ -12,29 +12,30 @@
...
-#include "dawn/Optimizer/Interval.h"
+#include "dawn/IIR/Interval.h"
 #include <list>
 
 namespace dawn {
+namespace iir {
 
 class MultiInterval {
-  std::vector<Interval> intervals_;
+  std::vector<iir::Interval> intervals_;
```
Comment (twicki) [inline src/dawn/IIR/MultiInterval.h]: again, namespace obsolete

File: src/dawn/IIR/IntervalAlgorithms.cpp [hunk 18]
```diff
@@ -24,23 +25,25 @@ MultiInterval substract(const Interval& int1, const Interval& int2) {
...
-                                (lowerI2InInterval ? int2.lowerOffset() - 1 : int1.upperOffset())}};
+  return MultiInterval{
+      iir::Interval{(upperI2InInterval ? int2.upperLevel() : int1.lowerLevel()),
+                    (lowerI2InInterval ? int2.lowerLevel() : int1.upperLevel()),
+                    (upperI2InInterval ? int2.upperOffset() + 1 : int1.lowerOffset()),
+                    (lowerI2InInterval ? int2.lowerOffset() - 1 : int1.upperOffset())}};
 }
 
-MultiInterval substract(const Interval& int1, const MultiInterval& int2) {
+MultiInterval substract(const iir::Interval& int1, const MultiInterval& int2) {
```
Comment (twicki) [inline src/dawn/IIR/IntervalAlgorithms.cpp]: again, namespace obsolete

**Comment by twicki** [comment]
I start checking the diff of the generated files now

File: src/dawn/IIR/CMakeLists.txt (hunk 13 above)
Comment (cosunae) [reply src/dawn/IIR/CMakeLists.txt]: not sure either, 
I think what is important is the depencies of the subproject, that we dont end up in circular dependencies. To me at the moment looks fine, not sure, but still IIR depends on Optimizer, we should finalize it later on, and make it totally independent of Optimizer

**Comment by twicki** [comment]
image

**Pull request merged**